### ISSUE TITLE
[RFC] Compatibility let+ syntax

### ISSUE THREAD

**Issue by diml** [ocaml/dune#1891]
OCaml 4.08 has support for user defined let operators, which supersedes our hand-written preprocessor for `let%map`. I propose to make it so that with 4.08 we use no preprocessing at all, and with OCaml < 4.08 we use a shim preprocessor to backport the new syntax.

## Spec

We add a new `compat` preprocessing specification:

```scheme
(library
 (name dune)
 ...
 (preprocess compat))
```

When using OCaml >= 4.08, this will be equivalent to `no_preprocessing`. When using OCaml < 4.08, this will be equivalent to `(action (run %{libexec:dune:compat-pp.exe}))` where `compat-pp.exe` is a preprocessor that translate let operators into valid pre-408 OCaml code.

## Implementation

With OCaml < 4.08, `compat-pp.exe` will be a preprocessor that translate the `let+` syntax as follow:

```ocaml
let ( let+ ) = <expr>
let ( and+ ) = <expr>
let+ p1 = e1 and+ p2 = e2 and+ p3 = e3 in e

-->

let ( let__plus__ ) = <expr>
let ( and__plus__ ) = <expr>
let__plus__ (and__plus__ (and__plus__ e1 e2) e3) (fun ((p1, p2), p3) -> e)
```

This preprocessor will reuse the OCaml lexer and parser as much as possible. To do so, we will simply add a wrapper on top of `Lexer.token` that will perform the following translations:

```
- LPAREN LET PLUS --> LPAREN (LIDENT "let__plus")
- LPAREN AND PLUS --> LPAREN (LIDENT "and__plus")
- LET PLUS --> LET LBRACKETAT (LIDENT "+") RBRACKET
- AND PLUS --> LET LBRACKETAT (LIDENT "+") RBRACKET
```

We will then map the resulting parsed AST and apply the relevant transformations when seeing `[@+]` attributes on value bindings. The preprocessor will output a marshaled AST.

Because this program will use the compiler libraries, it will be vulnerable to breaking changes to the compiler libraries. Given that these libraries are unstable, we will make sure that with OCaml >= 4.08 this program contains no code at all to avoid breakages with new versions of OCaml or OCaml trunk.

**Comment by emillon** [comment]
If I understand it correctly, this would allow not only dune itself to use the new syntax, but also any project built with dune?

If that's the case, I think it's neat, but `compat` is a bit vague. How about `let-syntax-compat` or something like this?

**Comment by diml** [comment]
Yep, since it is now a more polish feature we might as well provide it.

For name, I didn't put `let` in it so that we can extend it in the future. For instance in 4.09 there will be custom match operators as well and we might want to support that. What about `compat-syntax` or `syntax-shims`?

**Comment by jberdine** [comment]
As a point regarding generality, perhaps this processor could be a home for backward compatibility for e.g. https://github.com/ocaml/ocaml/pull/1958 by expanding `module M(_:S) =` to `module M = functor (_:S) ->` for ocaml versions < 4.08.

**Comment by diml** [comment]
Yep, that's what I was thinking. However, it might be hard to implement the shim in some cases.

**Comment by emillon** [comment]
OK, thanks for the explanations. I think it's great to have something like this. If there is a need for a wider scope (e.g. somebody wants a polyfill for `let exception in`), this could also be maintained separately.

**Comment by emillon** [comment]
That was addressed in #1899.
